lookout/style/format/rules.py

```diff
@@ -147,28 +276,29 @@ def predict(self, X: numpy.ndarray) -> numpy.ndarray:
         :param X: input features.
         :return: array of the same length as X with predictions.
         """
-        compiled, rules = self._cache
         prediction = numpy.zeros(len(X), dtype=int)
         for xi, x in enumerate(X):
-            ris = self._compute_triggered(compiled, rules, x)
+            ris = self._compute_triggered(self._compiled, self._rules, x)
             if len(ris) == 0:
                 # self.log.warning("no rule!")
                 continue
             if len(ris) > 1:
                 confs = numpy.zeros(len(ris), dtype=numpy.float32)
                 for i, ri in enumerate(ris):
-                    confs[i] = rules[ri].stats.conf
-                winner = rules[ris[numpy.argmax(confs)]].stats.cls
+                    confs[i] = self._rules[ri].stats.conf
```

Comment:
Please define `rules` in the outer scope, each lookup costs time

Reply:
ok, https://gist.github.com/zurk/fd214b7b500ef50df0b2577d2cd38a69